tasmota/berry/drivers/Core200S.be

```diff
@@ -0,0 +1,712 @@
+    static OFF = NightLightMode("off")
+    static LOW = NightLightMode("low")
+    static FULL = NightLightMode("full")
+    def init(value)
+        super(self).init({"off": 0, "low": 50, "full": 100}, value)
+    end
+end
+
+def modulo(num, mod)
+    return ((num % mod) + mod) % mod
+end
+
+def info_c2s(msg)
+    log("C2S: " + msg, 2)
+end
+
+def log_c2s(msg)
+    log("C2S: " + msg)
+end
+
+class Core200SSerial : serial
+    var msg_count
+    var read_buffer
+    var error_count
+    var expected_replies
+
+    def init()
+        super(self).init(16, 17, 115200)
+        self.reset_counters()
+        self.read_buffer = bytes("")
+    end
+
+    def deinit()
+        self.close()
+    end
+
+    static def calc_checksum(msg)
+        var checksum = 0xff
+        var i = 0
+        while i < msg.size()
+            if (i != 5)
+                checksum -= msg.get(i)
+                checksum = modulo(checksum, 256)
+            end
+            i += 1
+        end
+        return checksum
+    end
+
+    def send(msg)
+        self.send_nowait(msg)
+        tasmota.delay(100)
+        return self.read_msg()
+    end
+
+    def send_nowait(msg)
+        if msg.size() >= 2 && msg[1] == 0x22
+            self.expected_replies += 1
+        end
+
+        # increase and set message counter
+        self.msg_count += 1
+        msg[2] = self.msg_count
+
+        # insert checksum
+        msg[5] = self.calc_checksum(msg)
+
+        # send message
+        var i = 0
+        while i < msg.size()
+            self.write(msg[i])
+            i += 1
+        end
+
+        log_c2s("Send " + msg.tohex())
+    end
+
+    def available()
+        self.read_buffer += self.read()
+        return self.read_buffer.siez()
+    end
+
+    # read but don't remove from buffer
+    def read_ahead(num)
+        var content = self.read()
+        self.read_buffer += content
+
+        if self.read_buffer.size() >= num
+            return self.read_buffer[0..(num-1)]
+        end
+
+        return nil
+    end
+
+    # read an remove from buffer
+    def consume(num)
+        var content = self.read_ahead(num)
+        if content != nil
+            self.read_buffer = self.read_buffer[num..]
+        end
+        return content
+    end
+
+    def read_msg(allow_no_msg)
+        var max_read_attempts = 10
+        var header = self.read_ahead(6)
+        if header == nil
+            header = bytes("0000")
+        end
+        while header[0] != 0xa5 || (header[1] != 0x12 && header[1] != 0x22)
+            if header != bytes("0000")
+                self.consume(1)
+            else
+                max_read_attempts -= 1
+                if max_read_attempts <= 0
+                    if !allow_no_msg
+                        info_c2s("Too many read attempts (couldn't read message header)")
+                        self.error_count += 1
+                    end
+                    return bytes("")
+                end
+                tasmota.delay(100)
```

Comment:
Remove delay

Reply:
we are in the middle of reading a message here. Removing this delay can lead to message not being available at the point in time we are in the loop. We can maybe reduce a bit, but right now there is synchronous communication happening, which means missing a message here can lead to incorrect results of commands. Or am I missing something here?

Reply:
Let's stick to this, but generally speaking try to avoid this and have some sort of minimalistic state machine instead. `tasmota.set_timer()` is an easy way to launch code in the future, very similar to javascript